### ISSUE TITLE
[float8] two typo bugs in `float8_transpose` axiswise-dim remap and `Float8LinearConfig.__post_init__` operand-precision assertion

### ISSUE THREAD

**Issue by GotFusion** [pytorch/ao#4443]
Filing as a tracking issue for two independent typo-class bugs in `torchao/float8/`. Each has its own one-line fix already in flight (see below) — bundling them here so the bugfix story is documented in one place.

## Action items

- [ ] #4038 — Bug 1 (`float8_transpose` axiswise-dim remap)
- [ ] #4115 — Bug 2 (`Float8LinearConfig.__post_init__` operand-precision assertion)

This issue can be closed once **both** PRs land. I'm deliberately not using `Fixes #4443` in either PR body so that merging the first PR doesn't auto-close the issue while the second is still open.

---

## Bug 1 — `==` instead of `=` in `float8_transpose` axiswise dim remap

**Location**: [`torchao/float8/float8_ops.py:174-179`](https://github.com/pytorch/ao/blob/main/torchao/float8/float8_ops.py#L174-L179) (current `main`)

```python
new_axiswise_dim = old_axiswise_dim
if old_axiswise_dim is not None:
    if old_axiswise_dim == 0:
        new_axiswise_dim == -1   # ← comparison, result discarded
    else:
        new_axiswise_dim == 0    # ← comparison, result discarded
```

**Effect**: `new_axiswise_dim` is never reassigned, so after `Float8TrainingTensor.transpose(0, 1)` or `.t()`, the `_axiswise_dim` attribute keeps the original value (e.g. stays `0` after a transpose that should flip it to `-1`). This silently produces wrong axis metadata that downstream matmul / scale-flattening code relies on. The companion `float8_view` (`float8_ops.py:213-239`) already does the correct remap, so the bug is just an outlier here.

**Repro**:
```python
import torch
from torchao.float8 import e4m3_dtype
from torchao.float8.float8_scaling_utils import hp_tensor_and_scale_to_float8
from torchao.float8.float8_utils import tensor_to_scale

a = torch.rand((16, 16), dtype=torch.bfloat16)
scale = tensor_to_scale(a, e4m3_dtype)
fp8 = hp_tensor_and_scale_to_float8(a, scale, e4m3_dtype, axiswise_dim=0)
assert fp8.t()._axiswise_dim == -1, (
    f"expected -1 after transpose, got {fp8.t()._axiswise_dim}"
)
# fails on current main with: expected -1 after transpose, got 0
```

**Fix**: PR #4038 — replace `==` with `=` (2 chars), plus extends `test_transpose` to assert `_axiswise_dim` on the output (the existing test only checked `_data` / `_scale`).

---

## Bug 2 — `cc1` read twice instead of `cc1` / `cc2` in operand-precision assertion

**Location**: [`torchao/float8/config.py:285-294`](https://github.com/pytorch/ao/blob/main/torchao/float8/config.py#L285-L294) (current `main`)

```python
for cc1, cc2, gemm_name in (
    (cc_i, cc_w, "output"),
    (cc_go, cc_w_gi, "grad_input"),
    (cc_i_gw, cc_go_gw, "grad_weight"),
):
    is_disabled_1 = cc1.scaling_type is ScalingType.DISABLED
    is_disabled_2 = cc1.scaling_type is ScalingType.DISABLED   # ← reads cc1 twice
    assert is_disabled_1 == is_disabled_2, (
        f"incompatible operand precision for {gemm_name}"
    )
```

**Effect**: The check is supposed to enforce "both operands of a GEMM must be either both DISABLED or both quantized" (per the comment immediately above). Because `cc2` is never read, the comparison is always `is_disabled_1 == is_disabled_1` → always `True` → assertion never fires. Misconfigured `Float8LinearConfig` (e.g. `cast_config_input=DISABLED` but `cast_config_weight=DYNAMIC`) gets silently accepted, then fails later in the matmul path with a much harder-to-debug error.

**Repro**:
```python
from torchao.float8.config import (
    CastConfig, Float8LinearConfig, ScalingType,
)
# This should raise AssertionError("incompatible operand precision for output")
# On current main: silently constructs the (illegal) config.
Float8LinearConfig(
    cast_config_input=CastConfig(scaling_type=ScalingType.DISABLED),
    cast_config_weight=CastConfig(scaling_type=ScalingType.DYNAMIC),
)
```

**Fix**: PR #4115 — change `cc1` to `cc2` on the second line (1 char), plus adds a regression test that constructs the mismatched config and verifies the assertion fires.

---

## Why one issue for two bugs

Both bugs share the same shape and same blast radius:

- Both are local typos in stable `torchao/float8/` code, not in prototype paths.
- Both silently produce wrong behavior (no exception, no warning) and only manifest much later as model accuracy or matmul-dispatch errors.
- Both have existing PRs with minimal fixes + tests that lock in the corrected behavior.

Grouping them in one issue makes it easier to track that the `torchao/float8/` validation/metadata layer got a tidy-up pass.

**Comment by GotFusion** [comment]
Splitting this umbrella into two focused issues so each one gets standard one-to-one auto-close behavior from its fixing PR:

- **#4444** ← fixed by #4038 (`==` → `=` in `float8_transpose`)
- **#4445** ← fixed by #4115 (`cc1` → `cc2` in `Float8LinearConfig.__post_init__`)

Closing as superseded. History (original repro + discussion) preserved on this issue for searchability.
